To implement the neural networks algorithm, we use the mini-case study customer churn dataset that we have previously pre-processed and saved in the ChurnFinal.csv file.  We use the following Python codes to import necessary libraries:

In [2]:
import warnings
warnings.filterwarnings('ignore')
import numpy as np
import pandas as pd

#Loading Dataset
data = pd.read_csv('ChurnFinal.csv')

df_inputs = pd.get_dummies(data[['Gender', 'Age', 'PostalCode', 'Cash', 'CreditCard', 
        'Cheque', 'SinceLastTrx', 'SqrtTotal', 'SqrtMax', 'SqrtMin']])
df_label = data['Churn']

Based on our problem and data understanding for the customer churn case study (refer to Week 2 content), the solution requires an approach to classify if a customer is going to churn. Thus, we indicate the target attribute as Churn in the data set. We use Gender, Age, PostalCode, Cash, CreditCard, Cheque, SinceLastTrx, SqrtTotal, SqrtMax, and SqrtMin attributes as inputs to predict the churn by assigning them to user-defined variables df_label and df_inputs respectively for later use.

Next, we spit the original data set to train and test the model, using the Python train_test_split () function with the following example codes:

In [3]:
from sklearn.model_selection import train_test_split
X_train,X_test,Y_train,Y_test = train_test_split(df_inputs, df_label, 
                    stratify=df_label, test_size=0.3, random_state=1) 

In the above codes, we split 80% of the data set for training the model and 20% (i.e., test_size=0.2) as a test set to assess the model.  The random state (random_state) is a seed to the random number generator to ensure numbers are generated in the same order. 

For a detailed explanation of the train_test_split() API parameters, refer to the official website https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html?highlight=train_test_split#sklearn.model_selection.train_test_split.

After splitting the data set into train and test sets, we define the neural network algorithm as the data modeling technique using MLPClassifier(), and train the decision tree using fit() function the train data set (X_train and Y_train) in  function:

In [4]:
# feature scaling
from sklearn.preprocessing import StandardScaler  
scaler = StandardScaler()  
scaler.fit(X_train)                     #fit only on training data  
X_train = scaler.transform(X_train)  
X_test = scaler.transform(X_test)       # apply same transformation to test data

from sklearn.neural_network import MLPClassifier 
nn = MLPClassifier(solver='lbfgs', alpha=1e-05 , activation = 'identity', random_state=1,
             hidden_layer_sizes=(200,6), learning_rate = 'adaptive', 
             learning_rate_init=0.0001, max_iter=500)

nn.fit(X_train, Y_train)

MLPClassifier(activation='identity', alpha=1e-05, hidden_layer_sizes=(200, 6),
              learning_rate='adaptive', learning_rate_init=0.0001, max_iter=500,
              random_state=1, solver='lbfgs')

In the above code, StandardScaler() function standardizes the data values into a standard format. StandardScaler standardizes an attribute by subtracting the mean and then scaling to unit variance. Unit variance means dividing all the values by the standard deviation. Further, we use fit_transform() along with the assigned object to transform and standardize the data. StandardScaler() further removes outliers (if they still exist in the data set after we pre-processed the data) since it involves the estimation of the empirical mean and standard deviation of each attribute.

For a detailed explanation of the MLPClassifier() API parameters, refer to the official website https://scikit-learn.org/stable/modules/generated/sklearn.neural_network.MLPClassifier.html 


To assess how well the decision tree model developed based on the criteria above, we use the test data (X_test) for the model to predict the churn outcomes using the predict() function. 

In [5]:
#Predict the response for test dataset
y_predict = nn.predict(X_test)    

from sklearn import metrics 
print("Model Accuracy     : ", round(metrics.accuracy_score(Y_test, y_predict),4))

Model Accuracy     :  0.7023


To assess the how well the classification outcomes, we can derive the model accuracy with the accuracy_score() function by comparing the predicted (i.e., y_predict) and actual (i.e., Y_test) churn outcomes. 

After running all the codes together given above, we obtain a model accuracy of 0.7023 printed on the console terminal as Moddel Accuracy :     0.7023.  This result indicates that the model has an accuracy of 70.23%, implying that 70.23% of the time it correctly classifies churn outcomes ('yes' or 'no').